DOC: update file path descriptions in IO docstrings

enricorotundo opened the pull request:
- [x] closes #24321
- [ ] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Validate Docstrings script output:

```
 ✔  python scripts/validate_docstrings.py pandas.io.parquet.read_parquet                                                                                                                                                       (pandas-dev)

################################################################################
################## Docstring (pandas.io.parquet.read_parquet) ##################
################################################################################

Load a parquet object from the file path, returning a DataFrame.

.. versionadded 0.21.0

Parameters
----------
path : various
    Either a path to a file (a :class:`python:str`,
    :class:`python:pathlib.Path`,
    or :class:`py:py._path.local.LocalPath`),
    URL (including http, ftp, and S3 locations), or any object
    with a ``read()`` method (such as an open file or
    :class:`~python:io.BytesIO`).
columns : list, default=None
    If not None, only these columns will be read from the file.

    .. versionadded 0.21.1
engine : {'auto', 'pyarrow', 'fastparquet'}, default 'auto'
    Parquet library to use. If 'auto', then the option
    ``io.parquet.engine`` is used. The default ``io.parquet.engine``
    behavior is to try 'pyarrow', falling back to 'fastparquet' if
    'pyarrow' is unavailable.
kwargs are passed to the engine

Returns
-------
DataFrame

################################################################################
################################## Validation ##################################
################################################################################

5 Errors found:
	Parameters {**kwargs} not documented
	Unknown parameters {kwargs are passed to the engine}
	Parameter "kwargs are passed to the engine" has no type
	Parameter "kwargs are passed to the engine" has no description
	Return value has no description
2 Warnings found:
	See Also section not found
	No examples section found
```

Commit:
included more supported types by read_parquet in its docstring

Commit:
fixed pep8 E501

xhochy commented on pandas/io/parquet.py:
```diff
@@ -260,8 +260,13 @@ def read_parquet(path, engine='auto', columns=None, **kwargs):
+        :class:`python:pathlib.Path`,
+        or :class:`py:py._path.local.LocalPath`),
+        URL (including http, ftp, and S3 locations), or any object
+        with a ``read()`` method (such as an open file or
```
``read()`` and ``seek()`` are required; I'd rather describe it as a file-like object instead of mentioning the explicit methods that need to be implemented.

enricorotundo replied:
Agree, file-like is more clear. Just for the record I took that description from https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html#basic, and replaced StringIO with ByteIO

jorisvandenbossche commented on pandas/io/parquet.py:
```diff
@@ -260,8 +260,13 @@ def read_parquet(path, engine='auto', columns=None, **kwargs):
-    path : string
-        File path
+    path : various
+        Either a path to a file (a :class:`python:str`,
```
can you remove the "python" in the reference? I don't think that it is needed to ensure a correct link) (same for the ones below)

WillAyd commented:
@enricorotundo can you merge master and address comments here?

Commit:
merged master

Commit:
addressed comments from #25164

enricorotundo commented:
* merged master
* addressed comments
* passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

Validate Docstrings script output:
```
 ✔  python scripts/validate_docstrings.py pandas.io.parquet.read_parquet                                                                                                                                                       (pandas-dev)

################################################################################
################## Docstring (pandas.io.parquet.read_parquet) ##################
################################################################################

Load a parquet object from the file path, returning a DataFrame.

.. versionadded 0.21.0

Parameters
----------
path : various
    Either a path to a file (a :class:`str`,
    :class:`pathlib.Path`,
    or :class:`py:py._path.local.LocalPath`),
    URL (including http, ftp, and S3 locations), or any file-like object.
columns : list, default=None
    If not None, only these columns will be read from the file.

    .. versionadded 0.21.1
engine : {'auto', 'pyarrow', 'fastparquet'}, default 'auto'
    Parquet library to use. If 'auto', then the option
    ``io.parquet.engine`` is used. The default ``io.parquet.engine``
    behavior is to try 'pyarrow', falling back to 'fastparquet' if
    'pyarrow' is unavailable.
columns : list, default=None
    If not None, only these columns will be read from the file.

    .. versionadded 0.21.1
**kwargs
    Any additional kwargs are passed to the engine.

Returns
-------
DataFrame

################################################################################
################################## Validation ##################################
################################################################################

2 Errors found:
	Wrong parameters order. Actual: ('path', 'engine', 'columns', '**kwargs'). Documented: ('path', 'columns', 'engine', '**kwargs')
	Return value has no description
2 Warnings found:
	See Also section not found
	No examples section found
```

jreback commented on pandas/io/parquet.py:
```diff
@@ -260,8 +260,15 @@ def read_parquet(path, engine='auto', columns=None, **kwargs):
-    path : string
-        File path
+    path : various
+        Either a path to a file (a :class:`str`,
```
can you see what we say in other read_* methods and see if you can match them up

enricorotundo replied:
I checked some: `read_hdf`, `read_json`, `read_csv`, `read_excel` and `read_feather`. They all seem different: 

* param  sometimes is `path`, in others `filepath_or_buffer` or `path_or_buf`
* param type is also different across methods. Probably `str, path object, URL or file-like object` would be the most complete and concise, but does it work for all methods?

enricorotundo replied:
Those differences confuse the user quite a bit. I issued this PR when I found out I could actually use URLs also with `read_parquet`, and not just with `read_csv`.

enricorotundo replied:
@jreback I could apply a custom variation of `str, path object, URL or file-like object` to each read_* method as a start. What about that?

Commit:
fixed valdiation string error

jreback commented:
yes that’s what i am suggesting - we weant them all to conform to similar to what u have done here

Commit:
Merge branch 'master' into read_parquet-docs

Commit:
updated filepath_or_buffer param type docstring to pandas.read_csv

Commit:
updated io param type docstring to pandas.read_excel

Commit:
updated `path` param type docstring of pandas.read_feather + added missing full stops

Commit:
updated `path_or_buf` param type docstring of pandas.read_hdf

Commit:
updated `io` param type docstring of pandas.read_html

Commit:
updated `path_or_buf` param type docstring of pandas.read_json

Commit:
updated `path_or_buf` param type docstring of pandas.read_msgpack

Commit:
updated `path` param type docstring of pandas.read_parquet

Commit:
updated `filepath_or_buffer` param type docstring of pandas.read_sas

Commit:
updated `filepath_or_buffer` param type docstring of pandas.read_stata

enricorotundo commented:
@jreback please check the latests commits. I assume reading from a URL work in every read_* method in this PR. Is that a valid assumption?

jreback commented:
> @jreback please check the latests commits. I assume reading from a URL works in every read_* method in this PR. Is that a valid assumption?

yes we use a common reading routine.

cc @jorisvandenbossche

enricorotundo commented:
Anything I can do for this PR at the moment?

jreback commented on pandas/io/json/json.py:
```diff
@@ -235,7 +235,8 @@ def read_json(path_or_buf=None, orient=None, typ='frame', dtype=None,
     Parameters
     ----------
-    path_or_buf : a valid JSON string or file-like, default: None
+    path_or_buf : str, path object, URL, a valid JSON str or file-like object,\
```
is this backslash needed here?

jorisvandenbossche replied:
It's because the line was getting too long. However, I think we should actually just remove the `default: None`, as I don't see how this is useful? (meaning, passing `None` will raise an error, as a proper value is expected. I am not sure why `path_or_buf` is not a required positional argument)

jorisvandenbossche replied:
Another remark: previously there was "a valid JSON string" in the explanation, which was actually useful. So this needs to be added back.

enricorotundo replied:
Fixed in 4b1171a

jreback reviewed (changes requested):
tiny comment

jorisvandenbossche reviewed (commented):
This might be for another PR, but, reading the diff here, I think it would be good to explain somewhere (eg in the io.rst user guide) in more detail what those options are.

Another general remark: you now listed "URL" explicitly in the types list, but should this URL not be a string or path object?

jreback commented:
can you merge master

Commit:
Merge branch 'master' into read_parquet-docs

Commit:
docstring of read_json - addressed #25164 reviewer comment

Commit:
re-added ExcelFile to read_excel io param type list

enricorotundo commented on pandas/io/excel/_base.py:
```diff
@@ -32,7 +32,7 @@
 Parameters
 ----------
-io : str, file descriptor, pathlib.Path, ExcelFile or xlrd.Book
+io : str, xlrd.Book, path object, URL or file-like object
```
`ExcelFile` was removed by mistake. I've re-added it in 743f2a6

enricorotundo commented:
> This might be for another PR, but, reading the diff here, I think it would be good to explain somewhere (eg in the io.rst user guide) in more detail what those options are.

Agreed, will be happy to issue another PR on that.

> Another general remark: you now listed "URL" explicitly in the types list, but should this URL not be a string or path object?

You're right, since in this PR I tried to replace generic names with python types (`string, buffer` -> `str, file-like object`) it'd make sense to stick with types. My understanding is that if [io.common. get_filepath_or_buffer](https://github.com/pandas-dev/pandas/blob/5c341dc136cae12bf23c1a4b8e34c9951828c0cf/pandas/io/common.py#L179) can parse it, then it's a valid string, so not all string al valid; how to make that clear?

enricorotundo commented:
Not sure if pipelines fail due to my changes, seem most errors relate to `Series`...

WillAyd commented:
Can you merge master? Fixed issues with docstring validation on master within the past few hours

Commit:
Merge branch 'master' into read_parquet-docs

Commit:
Merge branch 'master' into read_parquet-docs

Commit:
fixed error PR05: pandas.read_json: Parameter "path_or_buf" type should not finish with "."

jreback commented on pandas/io/packers.py:
```diff
@@ -160,14 +160,14 @@ def writer(fh):
     Parameters
     ----------
-    path_or_buf : string File path, BytesIO like or string
+    path_or_buf : str, path object, URL or file-like object
```
can u be more explicit about path object; essentially say what you are doing below when you reference LocalPath

jreback replied:
do this everywhere you say path object

enricorotundo replied:
What about *"If you want to pass in a path object, pandas accepts either pathlib.Path or py._path.local.LocalPath."* as in [pd.read_csv](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html)? 

Actually, we could first agree upon a paragraph per type, then place blocks in each `read_*`, in the same order as the types appear in `path_or_buf `. For instance:

1. `str, URL`: Any valid string path is acceptable. The string could be a URL. Valid URL schemes include http, ftp, s3, and file. For file URLs, a host is expected. A local file could be: file://localhost/path/to/table.csv.
2. `path object`: If you want to pass in a path object, pandas accepts either pathlib.Path or py._path.local.LocalPath.
3. `file-like object`: By file-like object, we refer to objects with a read() method, such as a file handler (e.g. via builtin open function) or StringIO.

Thus a `path_or_buf : str, path object, URL or file-like object` would have (1), (2) and (3) in the description.

jreback replied:
yes that is the ideal to have identially phrases; a way to do this would be to define these phrases in pandas.core.generic, in a shared_docs entry; then simply compose them.

enricorotundo replied:
Done this in 6a3704bccbcb34e51513c2d0836b028ef5f2f4bc 1864c183ddb0a410e64d26019e4f9d9d9f572840 and 2ec85f9f5d2f101ec71a249352dd7122bc741c84, just by replicating the same description over and over

enricorotundo replied:
@jreback can you check my comment above?

jreback commented:
can you merge master and update as above

Commit:
Merge branch 'master' into read_parquet-docs

WillAyd commented:
Removing from 0.25 as not a blocker, but @enricorotundo if you can merge master and address comments would still take this PR!

jorisvandenbossche commented:
@enricorotundo already merged master since the last time it was asked, so I think we should first see if we are actually OK with the current diff, or if there are outstanding change requests.

jorisvandenbossche commented on pandas/io/excel/_base.py:
```diff
@@ -33,7 +33,7 @@
 Parameters
 ----------
-io : str, file descriptor, pathlib.Path, ExcelFile or xlrd.Book
+io : str, ExcelFile, xlrd.Book, path object, URL or file-like object
```
strictly speaking, in practice the URL is also a str or path object, so not sure we should list it in the types. It is clearly mentioned in the description below.

enricorotundo replied:
addressed in 4eb3fa5

enricorotundo replied:
I did the same in 6a3704bccbcb34e51513c2d0836b028ef5f2f4bc  for all other read_* methods I touched

jorisvandenbossche reviewed (approved):
Added a minor comment. But overall this is a clear improvement, so I am fine with just merging this.

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into read_parquet-docs

Commit:
removed 'URL' from io type list in pd.io.excel.base

Commit:
removed 'URL' from io type list in various read_* methods + more uniform filepath_or_buffer description

pep8speaks commented:
Hello @enricorotundo! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:







* In the file [`pandas/io/json/json.py`](https://github.com/pandas-dev/pandas/blob/6a3704bccbcb34e51513c2d0836b028ef5f2f4bc/pandas/io/json/json.py):

> [Line 233:80](https://github.com/pandas-dev/pandas/blob/6a3704bccbcb34e51513c2d0836b028ef5f2f4bc/pandas/io/json/json.py#L233): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)

* In the file [`pandas/io/packers.py`](https://github.com/pandas-dev/pandas/blob/6a3704bccbcb34e51513c2d0836b028ef5f2f4bc/pandas/io/packers.py):

> [Line 149:80](https://github.com/pandas-dev/pandas/blob/6a3704bccbcb34e51513c2d0836b028ef5f2f4bc/pandas/io/packers.py#L149): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)

* In the file [`pandas/io/parquet.py`](https://github.com/pandas-dev/pandas/blob/6a3704bccbcb34e51513c2d0836b028ef5f2f4bc/pandas/io/parquet.py):

> [Line 233:80](https://github.com/pandas-dev/pandas/blob/6a3704bccbcb34e51513c2d0836b028ef5f2f4bc/pandas/io/parquet.py#L233): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)

* In the file [`pandas/io/parsers.py`](https://github.com/pandas-dev/pandas/blob/6a3704bccbcb34e51513c2d0836b028ef5f2f4bc/pandas/io/parsers.py):

> [Line 727:80](https://github.com/pandas-dev/pandas/blob/6a3704bccbcb34e51513c2d0836b028ef5f2f4bc/pandas/io/parsers.py#L727): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)



* In the file [`pandas/io/sas/sasreader.py`](https://github.com/pandas-dev/pandas/blob/6a3704bccbcb34e51513c2d0836b028ef5f2f4bc/pandas/io/sas/sasreader.py):

> [Line 17:80](https://github.com/pandas-dev/pandas/blob/6a3704bccbcb34e51513c2d0836b028ef5f2f4bc/pandas/io/sas/sasreader.py#L17): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)

Commit:
addressed pep8speaks remarks

Commit:
more uniform path description for read_feather

datapythonista reviewed (approved):
lgtm, thanks @enricorotundo

Commit:
Merge branch 'master' into read_parquet-docs

Commit:
merged master, fixed conflicts

enricorotundo commented:
@jorisvandenbossche I merged `upstream/master` but it seems to have an issue with code formatting.

<img width="1792" alt="Screen Shot 2019-07-08 at 21 11 55" src="https://user-images.githubusercontent.com/4340327/60836226-78deb000-a1c5-11e9-8b38-fd9e0e249ded.png">

datapythonista commented:
@enricorotundo the CI is green, not sure why the long line is not reported, but you can open a new issue for that, this seems unrelated to the changes here.

@jreback you requested changes, if you can take a look.

jreback commented on pandas/io/excel/_base.py:
```diff
@@ -41,10 +41,17 @@
+    expected. A local file could be: ``file://localhost/path/to/table.xlsx``.
+
+    If you want to pass in a path object, pandas accepts either
+    ``pathlib.Path`` or ``py._path.local.LocalPath``.
```
I wouldn't mention LocalPath anyl onger

enricorotundo replied:
Done in ad29cff4ea4ddae0e99803498033d03d179429f2

jreback reviewed (changes requested):
lgtm.  would just change to use pathlib.Path (and say pathlike) as that's the de-facto path object now

Commit:
Merge branch 'master' into read_parquet-docs

Commit:
replaced pathlib.Path and py._path.local.LocalPath with os.PathLike in #25164

enricorotundo commented:
> lgtm. would just change to use pathlib.Path (and say pathlike) as that's the de-facto path object now

Replaced in ad29cff4ea4ddae0e99803498033d03d179429f2

enricorotundo commented:
@jreback can you take a look?

TomAugspurger commented:
Thanks @enricorotundo. Nice PR!
